Fix part of #18305: Fixed issue with feature flag, angular related 404 error

mon4our opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #18305.
2. This PR does the following: Updates the feature flag method used for study guides to handle user groups and fixes the 404 error on the study guide viewer page.
3. (For bug-fixing PRs only) The original bug occurred because: The feature flag method's user parameter was set to None which skips the user groups check entirely. The 404 error was caused because the url for the study guide viewer page has url fragments. These types of urls need to be separately registered to prevent errors.

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct

<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

[Screencast from 2025-09-26 22-45-31.webm](https://github.com/user-attachments/assets/4f4be936-fd6d-465d-ae4c-d448904f2b68)


#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

mon4our committed:
Fixed issue with feature flag and angular related 404 error

mon4our committed:
Merge branch 'develop' into study-guides-feat-flag-fix

seanlip commented on core/controllers/subtopic_viewer.py:
```diff
@@ -99,7 +99,7 @@ def get(self, topic_name: str, subtopic_id: int) -> None:
         if feature_flag_services.is_feature_flag_enabled(
             feature_flag_list.FeatureNames
-            .SHOW_RESTRUCTURED_STUDY_GUIDES.value, None
+            .SHOW_RESTRUCTURED_STUDY_GUIDES.value, self.user_id
```
@mon4our We must think about how to avoid this error happening in the future. Could you please verify that all other calls to is_feature_flag_enabled() always pass a user_id, so that folks who copy an existing implementation never use None?

mon4our replied:
Done!

seanlip reviewed (approved):
Thanks @mon4our -- LGTM, one comment.

(I think this might need approval from @Nik-09 too)

mon4our committed:
Added user_id in feature flag check method wherever possible

mon4our committed:
Updated tests

mon4our committed:
Updated screenshot

mon4our committed:
Merge branch 'develop' into study-guides-feat-flag-fix

mon4our commented:
@kevintab95 @imchristie PTAL, thanks!

mon4our committed:
Updated screenshots

seanlip reviewed (changes requested):
Changing to a blocking review to ensure https://github.com/oppia/oppia/pull/23480#discussion_r2383023718 is addressed.

mon4our committed:
Dummy screenshots

mon4our commented:
@seanlip I've addressed your comment in this commit: https://github.com/oppia/oppia/pull/23480/commits/b78d4b7593a3fcc6525c7d49383418a016f99f24

PTAL, thanks!

seanlip reviewed (approved):
LGTM. Thanks!

mon4our committed:
Updated screenshots

mon4our committed:
Merge branch 'develop' into study-guides-feat-flag-fix

mon4our committed:
Updated screenshots and made acceptance test deterministic

mon4our committed:
Reverted acceptance test related commits
